src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{fabric::Network, prss::SpaceIndex},
+    protocol::{
+        context::ProtocolContext,
+        modulus_conversion::gen_random::{GenRandom, ReplicatedBinary},
+        reveal_additive_binary::RevealAdditiveBinary,
+        RecordId, Step,
+    },
+    secret_sharing::Replicated,
+};
+use futures::future::try_join_all;
+
+pub struct XorShares {
+    num_bits: u8,
+    packed_bits: u64,
+}
+
+pub struct ConvertShares {
+    input: XorShares,
+}
+
+///
+/// This is an implementation of
+/// Protocol 5.2 Modulus-conversion protocol from `Z_2^u` to `Z_p`
+/// from the paper <https://eprint.iacr.org/2018/387.pdf>
+///
+/// It works by generating two secret-sharings of a random number `r`,
+/// one in `Z_2`, the other in `Z_p`. The sharing in `Z_2` is subtracted
+/// from the input and the result is revealed.
+///
+/// If the revealed result is `0`, that indicates that `r` had the same value
+/// as the secret input, so the sharing in `Z_p` is returned.
+/// If the revealed result is a `1`, that indicate that `r` was different than
+/// the secret nput, so the sharing 1 - the sharing in `Z_p` is returned.
+impl ConvertShares {
+    #[allow(dead_code)]
+    pub fn new(input: XorShares) -> Self {
+        Self { input }
+    }
+
+    #[allow(dead_code)]
+    pub async fn execute<F: Field, S: Step + SpaceIndex, N: Network<S>>(
+        &self,
+        ctx: &ProtocolContext<'_, S, N>,
+        record_id: RecordId,
+        step0: S,
+        step1: S,
+        step2: S,
+        step3: S,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let prss = &ctx.participant[step0];
+        let (left, right) = prss.generate_values(record_id.into());
+
+        let futures = (0..self.input.num_bits).into_iter().map(|i| async move {
+            let inner_record_id = RecordId::from(
```

Comment:
This looks hacky to me - is it possible that the upcoming `Step` change fixes this? It still makes more sense to me to move bit number to the step - that avoids u32 overflow here (the consequence of it would be two conversions generating the same `inner_record_id` )

Reply:
Yeah, I wasn't sure what was the best way to use the infra here. I'm open to whatever you and @martinthomson think best here.

Honestly, there's no dependency between the record being processed and the production of this "double random". In fact, if we wanted to, we could save query-latency by just pre-generating a gigantic batch of "double randoms" and saving them to disk. We could run that at non-peak times. Then, when we need to process a query we could just take them from the bank.